src/scripts/sync_synonym.py

```diff
@@ -0,0 +1,147 @@
+from oaklib.implementations import SqlImplementation
+from oaklib.types import CURIE, PRED_CURIE
+
+HERE = Path(os.path.abspath(os.path.dirname(__file__)))
+SRC_DIR = HERE.parent
+PROJECT_ROOT = SRC_DIR.parent
+sys.path.insert(0, str(PROJECT_ROOT))
+from src.scripts.utils import PREFIX_MAP, get_owned_prefix_map
+
+
+def _get_synonyms(ids: List[CURIE], db: SqlImplementation) -> pd.DataFrame:
+    """Get synonym triples from sqlite DB"""
+    tups: Set[Tuple[CURIE, PRED_CURIE, str]] = set()
+    for _id in ids:
+        for pred, alias in sorted(db.alias_relationships(_id)):
+            # - filter all but synonyms: prior in pipeline have already converted other props such as 'label' to syns
+            # todo: is this the best filter? is there more that can appear that should be filtered out? Should instead
+            #  not filter by things we want out, but by the list of synonym preds we want in?
+            if pred == 'rdfs:label':
+                continue
+            tups.add((_id, pred, alias))
+    df = pd.DataFrame(tups, columns=['curie', 'synonym_scope', 'synonym'])
+    return df
+
+
+# TODO
+def sync_synonyms(
+    ontology_db_path: Union[Path, str], mondo_scopes_and_xrefs_path: Union[Path, str],
+    mondo_mappings_path: Union[Path, str], onto_config_path: Union[Path, str], outpath_added: Union[Path, str],
+    outpath_confirmed: Union[Path, str], outpath_deleted: Union[Path, str], outpath_updated: Union[Path, str]
+):
+    """Create outputs for syncing synonyms between Mondo and its sources."""
+    # Iterate over mappings
+    mappings_df = pd.read_csv(mondo_mappings_path, sep='\t', comment='#')
+    # todo: refactor filtering by prefixes
+    #  check if this pattern has been used elsewhere; I think it has. and abstract into func for these 3 lines; 'filter prefixes'
+    #  - OR use this?: get_all_owned_terms()
+    #    - it seems long and messy compared to these 3 lines, but is more comprehensive. I could make get_all_owned_terms() a simple 3 line func and then make a separate func for all edge cases?
+    # - filter by source
+    owned_prefix_map: PREFIX_MAP = get_owned_prefix_map(onto_config_path)
+    mappings_df['object_prefix'] = mappings_df['object_id'].apply(lambda x: x.split(':')[0])
+    mappings_df = mappings_df[mappings_df['object_prefix'].isin(owned_prefix_map.keys())]
+    del mappings_df['object_prefix']
+
+    # Query synonyms
+    # - Mondo
+    mondo_df = pd.read_csv(mondo_scopes_and_xrefs_path, sep='\t').rename(columns={'?dbXref': 'source_id'})
+    mondo_df = mondo_df.rename(columns=lambda x: x.lstrip('?'))
+    mondo_df['object_prefix'] = mondo_df['source_id'].apply(lambda x: x.split(':')[0])
+    mondo_df = mondo_df[mondo_df['object_prefix'].isin(owned_prefix_map.keys())]
+    del mondo_df['object_prefix']
+    mondo_df['unprefixed_source_id'] = mondo_df['source_id'].apply(lambda x: x.split(':')[1])
+    mondo_df = mondo_df[mondo_df['unprefixed_source_id'].str.isnumeric()]
```

Comment:
## Removal of non-source_id, source-based xrefs
- [ ] @matentzn @twhetzel Review & suggest changes if needed

I had to remove certain cases like `OMIM:genemap2`.

In Mondo, we have situations like this:
```obo
id: MONDO:0008021
synonym: "Lhermitte-Duclos syndrome" EXACT [OMIM:158350, OMIM:genemap2]
```

I have to filter these out.
It seemed like the simplest way to do this was `.isnumeric()`. That assumes that all of our sources have CURIES like PREFIX:#. If that isn't the case (e.g. like in the future if we allow for `MIM:PS#`), then those will get filtered out... I don't know if we ever expect to have such cases. Does this seem like a good way to do this?  If not, any suggestions?

Reply:
There should be a fixed list of these "other" xref annotations that could be excluded explicitly.

Reply:
That's the main alternative I was considering, with the disadvantage being that it would need to be upkept.

Reply:
I see, I thought that existed in the Mondo documentation but can not find a reference for this one so checking if the value is numeric sounds good.

Reply:
However, doing a quick look through Mondo there are some cases I don't think that will work:
`MONDO:8000018 synonym: "benign paroxysmal positional vertigo" EXACT [DOID:13941, ICD9CM:386.11]`

There are also cases where Mondo IDs (e.g. `MONDO:0700140`) have an OMIA source such as `synonym: "osteosarcoma, dog" EXACT [OMIA:001441-9615]`.

What has your analysis of the values for the synonym xrefs shown?

Reply:
I dont know the whole context of this but can you remind me which part of the task requires you to read the xrefs of the Mondo synonyms?

Reply:
@matentzn It is necessary as part of working with triples (MONDO_ID, SYNONYM_PRED, SOURCE_ID). 

Example (re-using the one from [this comment](https://github.com/monarch-initiative/mondo-ingest/pull/93#discussion_r1669477452)): (`MONDO_0009092`, `hasExactSynonym`, `OMIM:221770`)

Here's the structure I'm querying:
```owl
    <owl:Class rdf:about="http://purl.obolibrary.org/obo/MONDO_0009092">
        <oboInOwl:hasExactSynonym>presenile dementia with bone cysts</oboInOwl:hasExactSynonym>
        ...
    </owl:Class>
    <owl:Axiom>
        <owl:annotatedSource rdf:resource="http://purl.obolibrary.org/obo/MONDO_0009092"/>
        <owl:annotatedProperty rdf:resource="http://www.geneontology.org/formats/oboInOwl#hasExactSynonym"/>
        <owl:annotatedTarget>presenile dementia with bone cysts</owl:annotatedTarget>
        <oboInOwl:hasDbXref>DOID:0090112</oboInOwl:hasDbXref>
        <oboInOwl:hasDbXref>OMIM:221770</oboInOwl:hasDbXref>
    </owl:Axiom>
```

So we're querying `hasDbXref` on the axiom for the synonym, not on the class itself.

---

There's also [this thread](https://github.com/monarch-initiative/mondo-ingest/pull/93#discussion_r1669477452), related to xrefs directly on the class, but that's a separate issue.

Reply:
@twhetzel Great examples, thanks for looking into this. `.isnumeric()` won't work.

I thought up a few alternative solutions; I could use one of these patterns as a way to distinguish between source IDs and not. @matentzn @twhetzel Let me know if you come up with any other / better suggestions.

**ID patterns**
a. doesn't contain any alpha chars?
b. doesn't start with an alpha char?
c. doesn't end with an alpha char?
d. doesn't start and end with an alpha char?
**non-ID patterns**
e. ~doesn't contain a numeric char?~ _(invalidated by `OMIM:genemap2`)

Reply:
> What has your analysis of the values for the synonym xrefs shown?

@twhetzel I haven't done such an analysis. I'm not 100% sure I know what you mean. Do you mean in order to identify the full list of cases like `OMIM:genemap2`?

Reply:
Yes, you can find everything that is numeric and therefore everything that isn't and with a unique list of those we can sort out what pattern to use vs. working by trial and error.

Reply:
- [ ] Trish found that `OMIM:genemap2` is the only such case, so I will statically program to remove that instead of `.isnumeric()`.

Reply:
Trish found that `OMIM:genemap2` is the only case, so I have statically programmed this case to be removed.

@twhetzel It is done. Marking this resolved.